include/dsn/utility/metrics.h

```diff
@@ -0,0 +1,92 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+
+#include <dsn/utility/autoref_ptr.h>
+#include <dsn/utility/ports.h>
+
+// A metric library (for details pls see https://github.com/apache/incubator-pegasus/issues/922)
+// inspired by Kudu metrics (https://github.com/apache/kudu/blob/master/src/kudu/util/metrics.h).
+//
+//
+// Example of defining and instantiating a metric entity
+// -----------------------------------------------------
+// Define an entity type at the top of your .cpp file (not within any namespace):
+// METRIC_DEFINE_entity(my_entity);
+//
+// To use the entity type, declare it at the top of any .h/.cpp file (not within any namespace):
+// METRIC_DECLARE_entity(my_entity);
+//
+// Instantiating the entity in whatever class represents the entity:
+// entity_instance = METRIC_ENTITY_my_entity.instantiate(my_entity_id, ...);
+
+// Define a new entity type.
+#define METRIC_DEFINE_entity(name) ::dsn::metric_entity_prototype METRIC_ENTITY_##name(#name)
```

Comment:
```suggestion
#define METRIC_DEFINE_entity(name) dsn::metric_entity_prototype METRIC_ENTITY_##name(#name)
```

Reply:
The prepended scope resolution operator can ensure that `::dsn::metric_entity_prototype` be resolved from global namespace even if  `METRIC_DEFINE_entity` is used within a namespace. Thus I think we can keep it.